APIv4 ConformanceTest - Demonstrate entity APIs which fail to emit `hook_civicrm_post(delete)`

totten opened the pull request:
Overview
----------------------------------------

Add assertions to APIv4's `ConformanceTest` to determine whether `FooEntity::delete()` emits `hook_civicrm_post('delete', 'FooEntity')`.

ping @colemanw

Before
----------------------------------------

Some entities emit `hook_civicrm_post('delete'...)`, and some don't... and it's hard to be sure which is which.

After
----------------------------------------

Some entities emit `hook_civicrm_post('delete'...)`, and some don't... and the test tells us which is which.

Comments
----------------------------------------

As discussed in https://github.com/civicrm/civicrm-core/pull/22199#issuecomment-984238304, the event `hook_civicrm_post('delete'...)` becomes more important in the current `master` -- if this event is not supported, then the managed-entities cannot be cleaned up.

The list of non-conformant entities is fairly long: `CaseType`, `Contact`, `ContactType`, `FinancialAccount`, `FinancialType`, `LocationType`, `MembershipStatus`, `MembershipType`, `MessageTemplate`, `OptionGroup`, `PaymentProcessor`, `PaymentProcessorType`, `RelationshipType`, `WordReplacement`.

(Note: `Contact` is probably OK, but it fails the test because it uses alternative names like `Individual`.)

totten commented:
@colemanw Wonder if it might be easier to switch `ManagedEntities` to listen for `civi.dao.postDelete` instead of `hook_civicrm_post(delete)`? OTOH, I guess we don't really need full compliance - we only need compliance for entities that use `ManagedEntity` trait?

colemanw commented:
IMO its pretty ridiculous that so many entities are non-compliant, and speaks to how these BAOs were originally written (by hand, with some loose conventions but no conformance tests).

I think this test is great, we should fix the DAOs to comply, and also add tests for 'pre' and 'post' hooks for every crud action.

FYI, the direction I'm trying to take our DAOs is to use generic `create` and `delete` functions that rely on late-static-binding and hook listeners, so adding this test can help nudge us further along that direction.

totten commented:
A couple updates:

1. Added a carve-out for the pre-existing quirk where `Contact::delete` is emitted as `hook_post(delete, Individual, ...)`.
2. Added assertion about `hook_pre(delete)` that works the same as `hook_post(delete)`.
3. Extracted a helper function for capturing the log. Should be add more assertions.

colemanw commented:
I've fixed all 13 non-compliant entities: #22207

totten commented:
jenkins, test this please
